api4/websocket.go

```diff
@@ -31,6 +38,46 @@ func connectWebSocket(c *Context, w http.ResponseWriter, r *http.Request) {
 
 	wc := c.App.NewWebConn(ws, *c.App.Session(), c.App.T, "")
 
+	if *c.App.Config().ServiceSettings.EnableReliableWebSockets {
+		connID := r.URL.Query().Get(connectionIDParam)
+		if connID == "" {
+			// If not present, we assume client is not capable yet, or it's a fresh connection.
+			// We just create a new ID.
+			connID = model.NewId()
+		} else {
+			// If present, we check if it's present in the connection manager.
+			// TODO: the connection manager internally should forward the request
+			// to the cluster if it does not have it.
+			//
+			// If the connection is not present, then we assume either timeout,
+			// or server restart. In that case, we set a new one.
+			//
+			// Now we get the sequence number
+			seqVal := r.URL.Query().Get(sequenceNumberParam)
+			if seqVal == "" {
+				// Sequence_number must be sent with connection id.
+				// A client must be either non-compliant or fully compliant.
+				mlog.Error("Sequence number not present in websocket request")
+				wc.WebSocket.Close()
+				return
+			}
+			seq, err := strconv.Atoi(seqVal)
+			if err != nil {
+				mlog.Error("Invalid sequence number set in query param",
+					mlog.String("query", seqVal),
+					mlog.Err(err))
+				wc.WebSocket.Close()
+				return
+			}
+			wc.Sequence = int64(seq)
+			// Now if there have been past entries to be back-filled, we do it.
+			// First we find the right sequence number point.
+			// We start consuming from dead queue first, and then move to active queue
+		}
+		// In case of fresh connection id, sequence number is already zero.
+		wc.SetConnectionID(connID)
```

Comment:
Should we do some basic validation before setting these values? I am thinking the sequence should be a positive number and the ID should be our usual 26 alphanumeric string.

Reply:
Makes sense. :+1: